### ISSUE TITLE
Build Issue

### ISSUE THREAD

**Issue by Alexander9673** [neon-bindings/neon#589]
I am on a 64 bit Windows 10
using node v14.4

when i build, i keep getting this
```
Caused by:
  process didn't exit successfully: `C:\Users\alexa\Desktop\Projects\rs\core\native\target\debug\build\neon-sys-605efa0192ecf482\build-script-build` (exit code: 1)
--- stdout
'Skipping node-gyp installation as part of npm install.'
added 96 packages from 70 contributors and audited 96 packages in 26.019s
found 7 vulnerabilities (3 low, 1 moderate, 3 high)
  run `npm audit fix` to fix them, or `npm audit` for details
cargo:node_arch=x64
cargo:node_root_dir=C:\\Users\\alexa\\AppData\\Local\\node-gyp\\Cache\\14.4.0
cargo:node_lib_file=C:\\\\Users\\\\alexa\\\\AppData\\\\Local\\\\node-gyp\\\\Cache\\\\14.4.0\\\\<(target_arch)\\\\node.lib

> @ build-debug C:\Users\alexa\Desktop\Projects\rs\core\native\target\debug\build\neon-sys-90e5d8ebd408e3dd\out\native
> node-gyp build --debug


C:\Users\alexa\Desktop\Projects\rs\core\native\target\debug\build\neon-sys-90e5d8ebd408e3dd\out\native>if not defined npm_config_node_gyp (node "C:\Users\alexa\AppData\Roaming\npm\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" build --debug )  else (node "C:\Users\alexa\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js" build --debug )
Building the projects in this solution one at a time. To enable parallel build, please add the "-m" switch.
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(514,5): error MSB6001: Invalid command line switch for "CL.exe". System.ArgumentException: Item has already been added. Key in dictionary: 'NPM_PACKAGE_MAIN'  Key being added: 'npm_package_main' [C:\Users\alexa\Desktop\Projects\rs\core\native\target\debug\build\neon-sys-90e5d8ebd408e3dd\out\native\build\neon.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(514,5): error MSB6001:    at System.Collections.Hashtable.Insert(Object key, Object nvalue, Boolean add) [C:\Users\alexa\Desktop\Projects\rs\core\native\target\debug\build\neon-sys-90e5d8ebd408e3dd\out\native\build\neon.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(514,5): error MSB6001:    at System.Collections.Hashtable.Add(Object key, Object value) [C:\Users\alexa\Desktop\Projects\rs\core\native\target\debug\build\neon-sys-90e5d8ebd408e3dd\out\native\build\neon.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(514,5): error MSB6001:    at System.Collections.Specialized.StringDictionaryWithComparer.Add(String key, String value) [C:\Users\alexa\Desktop\Projects\rs\core\native\target\debug\build\neon-sys-90e5d8ebd408e3dd\out\native\build\neon.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(514,5): error MSB6001:    at System.Diagnostics.ProcessStartInfo.get_EnvironmentVariables() [C:\Users\alexa\Desktop\Projects\rs\core\native\target\debug\build\neon-sys-90e5d8ebd408e3dd\out\native\build\neon.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(514,5): error MSB6001:    at Microsoft.Build.Utilities.ToolTask.GetProcessStartInfo(String pathToTool, String commandLineCommands, String responseFileSwitch) [C:\Users\alexa\Desktop\Projects\rs\core\native\target\debug\build\neon-sys-90e5d8ebd408e3dd\out\native\build\neon.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(514,5): error MSB6001:    at Microsoft.Build.Utilities.ToolTask.ExecuteTool(String pathToTool, String responseFileCommands, String commandLineCommands) [C:\Users\alexa\Desktop\Projects\rs\core\native\target\debug\build\neon-sys-90e5d8ebd408e3dd\out\native\build\neon.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(514,5): error MSB6001:    at Microsoft.Build.CPPTasks.TrackedVCToolTask.TrackerExecuteTool(String pathToTool, String responseFileCommands, String commandLineCommands) [C:\Users\alexa\Desktop\Projects\rs\core\native\target\debug\build\neon-sys-90e5d8ebd408e3dd\out\native\build\neon.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(514,5): error MSB6001:    at Microsoft.Build.CPPTasks.TrackedVCToolTask.ExecuteTool(String pathToTool, String responseFileCommands, String commandLineCommands) [C:\Users\alexa\Desktop\Projects\rs\core\native\target\debug\build\neon-sys-90e5d8ebd408e3dd\out\native\build\neon.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(514,5): error MSB6001:    at Microsoft.Build.Utilities.ToolTask.Execute() [C:\Users\alexa\Desktop\Projects\rs\core\native\target\debug\build\neon-sys-90e5d8ebd408e3dd\out\native\build\neon.vcxproj]
TARGET = Some("x86_64-pc-windows-msvc")
HOST = Some("x86_64-pc-windows-msvc")
AR_x86_64-pc-windows-msvc = None
AR_x86_64_pc_windows_msvc = None
HOST_AR = None
AR = None
running: "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.26.28801\\bin\\HostX64\\x64\\lib.exe" "-out:C:\\Users\\alexa\\Desktop\\Projects\\rs\\core\\native\\target\\debug\\build\\neon-sys-90e5d8ebd408e3dd\\out\\libneon.a" "-nologo" "C:\\Users\\alexa\\Desktop\\Projects\\rs\\core\\native\\target\\debug\\build\\neon-sys-90e5d8ebd408e3dd\\out\\native\\build\\Debug\\obj\\neon\\neon.obj"
LINK : fatal error LNK1181: cannot open input file 'C:\Users\alexa\Desktop\Projects\rs\core\native\target\debug\build\neon-sys-90e5d8ebd408e3dd\out\native\build\Debug\obj\neon\neon.obj'
exit code: 1181

--- stderr
gyp ERR! build error
gyp ERR! stack Error: `C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin\MSBuild.exe` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onExit (C:\Users\alexa\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\build.js:194:23)
gyp ERR! stack     at ChildProcess.emit (events.js:315:20)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:276:12)
gyp ERR! System Windows_NT 10.0.19041
gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\alexa\\AppData\\Roaming\\npm\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "build" "--debug"
gyp ERR! cwd C:\Users\alexa\Desktop\Projects\rs\core\native\target\debug\build\neon-sys-90e5d8ebd408e3dd\out\native
gyp ERR! node -v v14.4.0
gyp ERR! node-gyp -v v5.1.0
gyp ERR! not ok
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @ build-debug: `node-gyp build --debug`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the @ build-debug script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\alexa\AppData\Roaming\npm-cache\_logs\2020-08-14T04_53_37_132Z-debug.log


error occurred: Command "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.26.28801\\bin\\HostX64\\x64\\lib.exe" "-out:C:\\Users\\alexa\\Desktop\\Projects\\rs\\core\\native\\target\\debug\\build\\neon-sys-90e5d8ebd408e3dd\\out\\libneon.a" "-nologo" "C:\\Users\\alexa\\Desktop\\Projects\\rs\\core\\native\\target\\debug\\build\\neon-sys-90e5d8ebd408e3dd\\out\\native\\build\\Debug\\obj\\neon\\neon.obj" with args "lib.exe" did not execute successfully (status code exit code: 1181).



warning: build failed, waiting for other jobs to finish...
error: build failed
neon ERR! cargo build failed

Error: cargo build failed
    at Target.<anonymous> (C:\Users\alexa\Desktop\Projects\rs\core\node_modules\neon-cli\lib\target.js:121:35)
    at step (C:\Users\alexa\Desktop\Projects\rs\core\node_modules\neon-cli\lib\target.js:32:23)
    at Object.next (C:\Users\alexa\Desktop\Projects\rs\core\node_modules\neon-cli\lib\target.js:13:53)
    at fulfilled (C:\Users\alexa\Desktop\Projects\rs\core\node_modules\neon-cli\lib\target.js:4:58)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! core@0.1.0 install: `neon build`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the core@0.1.0 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\alexa\AppData\Roaming\npm-cache\_logs\2020-08-14T04_53_38_088Z-debug.log
```

**Comment by kjvalencik** [comment]
I haven't seen this one before. It seems like Visual Studio doesn't like `npm_package_main` being set multiple times as an environment variable. Can you try removing the `install` script from `package.json` and manually running `neon build --release`?

**Comment by Alexander9673** [comment]
> I haven't seen this one before. It seems like Visual Studio doesn't like `npm_package_main` being set multiple times as an environment variable. Can you try removing the `install` script from `package.json` and manually running `neon build --release`?

same results for `npm install`, `cargo build`, `neon build` and `neon build --release`

**Comment by Alexander9673** [comment]
Actually, i suddenly keep getting `neon ERR! spawn cargo ENOENT` with `neon build`.
But i alr installed it with `npm install neon-cli -g`

**Comment by kjvalencik** [comment]
@Alexander9673 This is the only thing that I could find that appears relevant. https://github.com/dotnet/runtime/issues/26471

The stack showing appears like it's dying before it even attempts to build `neon-sys`:
```
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(514,5): error MSB6001: Invalid command line switch for "CL.exe". System.ArgumentException: Item has already been added. Key in dictionary: 'NPM_PACKAGE_MAIN'  Key being added: 'npm_package_main' [C:\Users\alexa\Desktop\Projects\rs\core\native\target\debug\build\neon-sys-90e5d8ebd408e3dd\out\native\build\neon.vcxproj]
```

**Comment by kjvalencik** [comment]
`cargo` can't be found on the path? That does kind of sound like an issue with environment variables being corrupted.

**Comment by Alexander9673** [comment]
> `cargo` can't be found on the path? That does kind of sound like an issue with environment variables being corrupted.

but i can run `cargo` properly?

**Comment by adumbidiot** [comment]
I ran into the same issue. `cargo` was definitely in my path. I managed to work around it somehow by purging `node_modules` and reinstalling everything. I used `neon-cli` though my `devDependencies` with an npm script instead of installing it globally. I'm running Windows 10. This is my `cargo -V`: `cargo 1.47.0-nightly (aa6872140 2020-07-23)`.  My `package.json` had `neon-cli` as `^0.3.3`.

**Comment by connor4312** [comment]
I hit this in CI just now https://github.com/connor4312/blake3/runs/1469332114

Here's the build scripts up to that point... it seems fairly innocuous https://github.com/connor4312/blake3/blob/cfa82d398be56bfaa5c56a540e4fcf196f0a8cc4/.github/workflows/build-neon.yml#L17-L27

**Comment by kjvalencik** [comment]
@connor4312 That looks like the known issue with node 15 + Windows + npm 7 (https://github.com/neon-bindings/neon/issues/642). We haven't been able to reproduce outside of CI in order to fix it.

Our current workaround is to downgrade to npm 6 on that specific combo (https://github.com/neon-bindings/neon/blob/28ae75160691b894fed78b3c15a57221b8261e13/.github/workflows/windows.yml#L33-L35).

**Comment by goto-bus-stop** [comment]
FWIW this specific error:
```
  LINK : fatal error LNK1181: cannot open input file 'D:\a\blake3\blake3\rs\native\target\x86_64-pc-windows-msvc\release\build\neon-sys-d5a548974dde87d5\out\native\build\Release\obj\neon\neon.obj'
```
should be addressed by Neon 0.5.3. It looks like that CI run is still using 0.3.3, which hasn't received a patch for improved Node 15 compatibility.

But you may still run into the issue that @kjvalencik mentioned after upgrading.

**Comment by Himself65** [comment]
I change node 15 to 14, and it builds success.

**Comment by killpowa** [comment]
any update on this issue? I am not able to compile my neon project using vs2019

**Comment by kjvalencik** [comment]
@killpowa There is no change. We haven't been able to reliably reproduce it. It seems like it's caused by a specific combination of VS2019, Node 15, and NPM 7. Downgrading NPM to 6 resolves the issue.

However, it's not all environments because my Windows VM is unaffected. I haven't been able to reproduce outside of CI.

Additionally the N-API backend is not impacted; only the legacy backend.

**Comment by kjvalencik** [comment]
This should be fixed in 0.10 by https://github.com/neon-bindings/neon/commit/62a5ffc1e606ad906e4ca350bbae25ba28a55fd0
